napari/plugins/hook_specifications.py

```diff
@@ -373,45 +373,69 @@ def napari_experimental_provide_function_widget() -> Union[
     AugmentedWidget, List[AugmentedWidget]
 ]:
-    """Provide QWidget classes to be instantiated and docked on the viewer.
+    """Provide functions that return widgets to be docked in the viewer.
 
     This hook specification is marked as experimental as the API or how the
     returned value is handled may change here more frequently then the
     rest of the codebase.
 
     Returns
     -------
-    dock_widget(s) : QWidget class or list of QWidget classes
-        Implementations should return either QWidget classes (one or a list),
-        or tuple(s) containing QWidget classes as well as a dictionary
-        containing keyword arguments for
-        :meth:`napari.qt.Window.add_dock_widget` (though note that the
-        ``shortcut=`` keyword is not yet supported).
+    resul : callable or tuple or list of callables or list of tuples
+        A "callable" in this context is a classe or functions that, when
+        called, returns an instance of either a
+        :class:`~qtpy.QtWidgets.QWidget` or a
+        :class:`~magicgui.widgets.FunctionGui`.
+
+        Implementations of this hook specification must return a callable, or a
+        tuple of ``(callable, dict)``, where the dict contains keyword
+        arguments for :meth:`napari.qt.Window.add_dock_widget`. (note, however,
+        that ``shortcut=`` keyword is not yet supported).
+
+        Implementations may also return a list, in which each item must be a
+        callable or ``(callable, dict)`` tuple.
 
     Examples
     --------
+    An example with a QtWidget:
+
     >>> from qtpy.QtWidgets import QWidget
     >>> from napari_plugin_engine import napari_hook_implementation
     >>>
     >>> class MyWidget(QWidget):
-    >>>     def __init__(self, napari_viewer):
-    >>>         self.viewer = napari_viewer
-    >>>         super().__init__()
-    >>>
-    >>>         # initialize layout
-    >>>         layout = QGridLayout()
+    ...     def __init__(self, napari_viewer):
+    ...         self.viewer = napari_viewer
+    ...         super().__init__()
+    ...
+    ...         # initialize layout
+    ...         layout = QGridLayout()
+    ...
+    ...         # add a button
+    ...         btn = QPushButton('Click me!', self)
+    ...         def trigger():
+    ...             print("napari has", len(napari_viewer.layers), "layers")
+    ...         btn.clicked.connect(trigger)
+    ...         layout.addWidget(btn)
+    ...
+    ...         # activate layout
+    ...         self.setLayout(layout)
     >>>
-    >>>         # add a button
-    >>>         btn = QPushButton('Click me!', self)
-    >>>         def trigger():
-    >>>             print("napari has", len(napari_viewer.layers), "layers")
-    >>>         btn.clicked.connect(trigger)
-    >>>         layout.addWidget(btn)
+    >>> @napari_hook_implementation
+    >>> def napari_experimental_provide_dock_widget():
+    ...     return MyWidget
+
+    An example using magicgui:
```

Comment:
> If people use the @magicgui decorator that will work too right? 

nope, that's the motivation for this whole factory thing.  to summarize the main points here:

- `viewer.window.add_dock_widget` expects a widget _instance_ (be it a `QWidget` instance or `FunctionGui` instance)
- the `provide_dock_widget` hookspec expects a widget _class_ (or, more generally, a callable that returns a widget instance).
- `magicgui(my_function)` returns a `FunctionGui` _instance_ and is therefore suited to `viewer.window.add_dock_widget` (but _not_ to `provide_dock_widget`, which is what @jni was trying to do with affinder)
- `magic_factory(my_function)` returns a _callable_ that returns a `FunctionGui` instance, and is therefore suited to `provide_dock_widget` (but not to `viewer.window.add_dock_widget`).

That's it, the `factory` returns something that behaves like a `class`, the original decorator returns something that behaves like an instance.

Reply:
> Do you want to add a note here about using the magic_factory decorator/ possibly with link to documentation as this will be first time people see it

sure, but still waiting for review on that PR, so no documentation to link to yet.

Reply:
Wow, love it!!! Yes, it all comes back now, but there's nothing quite like seeing the four bullet points all next to each other. We definitely want to make sure that text is easily discoverable to people as it will be confusing otherwise. Philosophically it makes sense to me that you are "adding" an instance vs "providing" a callable (I know this may not apply to all places where we use these works) but it does help me understand the difference between the window method and the hookspec

I can give the magic_factory PR some review now